### DISCUSSION TITLE
Fun vs Function & Subscript vs Sub

### DISCUSSION THREAD

**Discussion by sparky4pro** [hylo-lang/hylo#922] [Language design]
It's not really an idea, more an observation:

Please do not mix styles and keep using `fun` and `subscript`.
Either use `fun` and `sub`, or `function` and `subscript`.

BTW: 
If you decide on using `sub` you would make many former VB programmers happy :P

Thanks for reading.

**Comment by kyouko-taiga** [comment]
I don't have a particularly strong objection so it is something we can consider. How do others feel about such a change?

**Reply by dabrahams** [reply]
Not into it. I appreciate the complaint but I see no obvious good fixes. Fun is precedented as an abbrev for function but sub is just a prefix that attaches to thousands of words and could mean anything. And "function" would make lambdas even more verbose than they already are.

**Reply by kyouko-taiga** [reply]
Without wanting to push in any particular direction (really), I only want to point at that `sub` would probably not be confused as the prefix of any other word in a language where subscripts are an extremely common construction. The situation is different than if the abbreviation was used in one single type of the standard library, for example.

`subscript` is not a very common programming construct, so I do not expect people coming from say Python/JS to immediately recognize what it means. As such, I don't think seeing `sub` instead of `subscript` would dramatically worsen their first contact with the language.

**Reply by dabrahams** [reply]
Every language with arrays has something called a subscript, whether users can declare them or not

**Reply by sparky4pro** [reply]
There are 1000th of different languages out there, do you want to avoid every incarnation of `sub`?

As an outsider that dives into Hylo I look at it from this point of view: how logical is the new (to me) syntax laid out?
If there is only one keyword in Hylo that starts with `sub...` then (in my view) it's no problem to call `subscript` `sub`.
For an outsider coming to Hylo `fun` and `sub` makes a lot of sense.

In Odin `::` means something totally different than in C++; this didn't hinder me to write code in both languages...

BTW: it was not a complaint, it was a suggestion. I'm not a contributor, so I have no right to complain or to demand anything.

**Reply by dabrahams** [reply]
Sorry @sparkylein, I didn't mean to disparage your suggestion by calling it a complaint.  I don't understand this remark:

> There are 1000th of different languages out there, do you want to avoid every incarnation of `sub`?

My point was that when the language manual-or-whatever for 99% of these languages talks about expressions of the form `a[i]` they're going to use the word “subscript” (note I didn't write ”`subscript`”).  That's a word in common use for the general idea even if a language doesn't support declaring new subscripts, so it communicates a lot more than `sub`, which can easily be interpreted as meaning “substitution,” “subject,” …etc.  We're not going to have all introducers be three letters—we have `type` and `typealias` and `trait`—so I think creating consistency in this corner of the language is of questionable value.

Note also that we have a lot of experience with Swift, which uses `func` and `subscript`, and the difference in length has not been seen to be an impediment to anyone learning the language.

**Reply by sparky4pro** [reply]
> I don't understand this remark:

> > There are 1000th of different languages out there, do you want to avoid every incarnation of sub?

Well, this was a reply to:

> Every language with arrays has something called a subscript, whether users can declare them or not

I got the impression that you cared if other languages have something called `subscript`; that's all.
Thanks for reading.

**Reply by dabrahams** [reply]
I know what it was a reply to, but I still don't understand it.  If you don't think that matters, you don't need to explain, but I wanted to make sure I gave you a chance to be fully understood.

**Reply by sparky4pro** [reply]
Understood.
Thank you.

**Reply by RMJay** [reply]
I like `fun` because it is short enough to be used as closure syntax and so there can be consistency across all kinds of function declarations. I must admit I have wondered why `subscript` wasn't just `lense`?

**Reply by kyouko-taiga** [reply]
`lens` is a very interesting suggestion indeed. I think `subscript` makes sense for Hylo because it also relates to the way people apply this construct at call site.

**Reply by RMJay** [reply]
`lens` would be awesome. I think more teachable too. It highlights that it is more of a special concept, which it is in Hylo. The term `subscript` comes from linear algebra, `lens` is coming from computer science.

**Reply by dabrahams** [reply]
Computed properties are every bit as much lenses as subscripts are, but they are not declared with this keyword.  *Lens* is a fine term for the notional union of these two syntactic language features, `property` and `subscript`.  I agree with Dimi that naming these things for the syntactic effect they create (which should be familiar to *many* programmers from *many* languages) is going to be more relatable than to use an esoteric term that comes from functional programming and theoretical CS.

**Reply by dabrahams** [reply]
Subscripts may have originated in linear algebra (maybe) but they are a central concept in thousands of programming languages.
